chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,354 @@
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2019 Corporation for Digital Scholarship
+					 Vienna, Virginia, USA
+					 https://digitalscholar.org
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+const { COLUMNS: ITEMTREE_COLUMNS } = require("zotero/itemTreeColumns");
+
+/**
+ * @typedef {import("../itemTreeColumns.jsx").ItemTreeColumnOptions} ItemTreeColumnOptions
+ * @typedef {"dataKey" | "label" | "pluginID"} RequiredCustomColumnOptionKeys
+ * @typedef {Required<Pick<ItemTreeColumnOptions, RequiredCustomColumnOptionKeys>>} RequiredCustomColumnOptionsPartial
+ * @typedef {Omit<ItemTreeColumnOptions, RequiredCustomColumnOptionKeys>} CustomColumnOptionsPartial
+ * @typedef {RequiredCustomColumnOptionsPartial & CustomColumnOptionsPartial} ItemTreeCustomColumnOptions
+ * @typedef {Partial<Omit<ItemTreeCustomColumnOptions, "enabledTreeIDs">>} ItemTreeCustomColumnFilters
+ */
+
+class ItemTreeManager {
+	_observerAdded = false;
+
+	/** @type {Record<string, ItemTreeCustomColumnOptions}} */
+	_customColumns = {};
+
+	/**
+	 * Register a custom column. All registered columns must be valid, and must have a unique dataKey.
+	 * Although it's async, resolving does not promise the item trees are updated.
+	 * @param {ItemTreeCustomColumnOptions | ItemTreeCustomColumnOptions[]} options - An option or array of options to register
+	 * @returns {string | string[] | false} - The dataKey(s) of the added column(s) or false if no columns were added
+	 * @example
+	 * A minimal custom column:
+	 * ```js
+	 * // You can unregister the column later with Zotero.ItemTreeManager.unregisterColumns(registeredDataKey);
+	 * const registeredDataKey = await Zotero.ItemTreeManager.registerColumns(
+	 * {
+	 *     dataKey: 'rtitle',
+	 *     label: 'reversed title',
+	 *     pluginID: 'make-it-red@zotero.org', // Replace with your plugin ID
+	 *     dataProvider: (item, dataKey) => {
+	 *         return item.getField('title').split('').reverse().join('');
+	 *     },
+	 * });
+	 * ```
+	 * @example
+	 * A custom column using all available options.
+	 * Note that the column will only be shown in the main item tree.
+	 * ```js
+	 * const registeredDataKey = await Zotero.ItemTreeManager.registerColumns(
+	 * {
+	 *     dataKey: 'rtitle',
```

Comment:
Add a comment that this is prefixed by the plugin id?

Reply:
Done